security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisResultBuilder.java

```diff
@@ -26,179 +26,198 @@
-        }
-
-        currentBuilder.addViolation(violation);
-        return this;
-    }
-
-    public SecurityAnalysisResultBuilder setComputationOk(boolean computationOk) {
-        if (currentBuilder == null) {
-            throw new IllegalStateException("Cannot set computation status: currentBuilder is not set");
-        }
-
-        currentBuilder.setComputationOk(computationOk);
-        return this;
+    public PostContingencyResultBuilder contingency(Contingency contingency, SecurityAnalysisResultContext postContingencyResultContext) {
+        return new PostContingencyResultBuilder(contingency, postContingencyResultContext);
     }
 
     /**
      * Finalizes the result.
      * @return the N situation result builder
      */
     public SecurityAnalysisResult build() {
-        if (currentBuilder != null) {
-            throw new IllegalStateException("Cannot build the result: the currentBuilder is not terminated");
-        }
         if (preContingencyResult == null) {
             throw new IllegalStateException("Pre-contingency result is not yet defined, cannot build security analysis result.");
         }
 
         SecurityAnalysisResult res = new SecurityAnalysisResult(preContingencyResult, postContingencyResults);
         res.setNetworkMetadata(new NetworkMetadata(context.getNetwork()));
-        interceptors.forEach(i -> i.onSecurityAnalysisResult(context, res));
+        interceptors.forEach(i -> i.onSecurityAnalysisResult(res, context));
 
         return res;
     }
 
     /**
-     * Provides access to the security analysis running context to children classes.
-     * @return the security analysis running context.
+     * Base class for the pre and post contingency builders.
      */
-    protected RunningContext getContext() {
-        return context;
-    }
+    public abstract class AbstractLimitViolationsResultBuilder<B extends AbstractLimitViolationsResultBuilder<B>> {
 
-    private interface ResultBuilder {
+        protected boolean computationOk;
 
-        void setComputationOk(boolean computationOk);
+        protected final List<LimitViolation> violations = new ArrayList<>();
 
-        void addViolation(LimitViolation violation);
+        protected final SecurityAnalysisResultContext resultContext;
 
-    }
+        public B setComputationOk(boolean computationOk) {
+            this.computationOk = computationOk;
+            return (B) this;
+        }
 
-    private class LimitViolationsResultBuilder implements ResultBuilder {
+        /**
+         * Initiates a result builder with a {@link SecurityAnalysisResultContext}.
+         * @param resultContext The context would be used when creation result or as default context when a limit violation added.
+         */
+        public AbstractLimitViolationsResultBuilder(SecurityAnalysisResultContext resultContext) {
+            this.resultContext = resultContext;
```

Comment:
It seems that the context is never null. Is it a coincidence or a choice. I think we should assert that it's not null there and in the SecurityAnalysisResultBuilder constructor.

Reply:
Made it private and check non null on two usages